src/prefect/task_engine.py

```diff
@@ -409,6 +409,22 @@ def set_state(self, state: State, force: bool = False) -> State:
             self.task_run.state_id = new_state.id
             self.task_run.state_type = new_state.type
             self.task_run.state_name = new_state.name
+
+            if new_state.is_final():
+                if (
+                    isinstance(state.data, BaseResult)
+                    and state.data.has_cached_object()
+                ):
+                    # Avoid fetching the result unless it is cached, otherwise we defeat
+                    # the purpose of disabling `cache_result_in_memory`
+                    result = state.result(raise_on_failure=False, fetch=True)
```

Comment:
Fetch true here seems wrong given the pre condition above? Or am I misunderstanding

Reply:
to be totally honest with you I coped this out of `propose_state_sync`:

https://github.com/PrefectHQ/prefect/blob/f5d48925b2746013d715130ed6d7dc9c97a79748/src/prefect/utilities/engine.py#L500-L511

Reply:
I can update in both places if that is incorrect?

Reply:
i just looked at this code path, and wow that's some wild and poorly named code! Let's keep it as-is to preserve behavior and revisit this as tech debt soon